src/integrate/totalvi/script.py

```diff
@@ -0,0 +1,181 @@
+
+    vae_reference.save(par["reference_model_path"])
+
+    return vae_reference
+
+def is_retraining_model() -> bool:
+    """Decide, whether reference model should be trained. It happens when no model exists or force_retrain flag is on"""
+    
+    trained_model_exists = os.path.isdir(par["reference_model_path"]) and ("model.pt" in os.listdir(par["reference_model_path"]))
+    return not trained_model_exists or par["force_retrain"]
+
+
+def map_query_to_reference(mdata_reference: MuData, mdata_query: MuData, adata_query: AnnData) -> Tuple[scvi.model.TOTALVI, AnnData]:
+    """Build model on the provided reference if necessary, and map query to the reference"""
+
+    adata_reference: AnnData = convert_mudata_to_anndata(mdata_reference, rna_modality_key=par["reference_modality"], protein_modality_key=par["reference_proteins_key"],
+                                                         input_layer=par["input_layer"], hvg_var_key=par["var_input"])
+
+    scvi.model.TOTALVI.setup_anndata(
+        adata_reference,
+        batch_key=par["obs_batch"],
+        protein_expression_obsm_key=par["reference_proteins_key"]
+    )
+
+    if is_retraining_model():
+        vae_reference = build_reference_model(adata_reference, max_train_epochs=par["max_epochs"])
+    else:
+        vae_reference = scvi.model.TOTALVI.load(dir_path=par["reference_model_path"], adata=adata_reference)
+
+    adata_query: AnnData = align_proteins_names(adata_reference, mdata_query, adata_query, reference_proteins_key=par["reference_proteins_key"],
+                                       query_proteins_key=par["query_proteins_key"])
+
+    # Reorder genes and pad missing genes with 0s
+    scvi.model.TOTALVI.prepare_query_anndata(adata_query, vae_reference)
+
+    # Train the model for query
+    vae_query = scvi.model.TOTALVI.load_query_data(
+        adata_query,
+        vae_reference
+    )
+    vae_query.train(par["max_query_epochs"], plan_kwargs=dict(weight_decay=par["weight_decay"]))
+
+    return vae_query, adata_query
+
+def main():
+    logger = _setup_logger()
+
+    mdata_query = mudata.read(par["query"].strip())
+    adata_query = convert_mudata_to_anndata(mdata_query, rna_modality_key=par["query_modality"], protein_modality_key=par["query_proteins_key"],
+                                            input_layer=par["input_layer"], hvg_var_key=par["var_input"])
+
+    if par["reference"].endswith(".h5mu"):
+        logger.info("Reading reference")
+        mdata_reference = mudata.read(par["reference"].strip())
+
+        logger.info("Mapping query to the reference")
+        vae_query, adata_query = map_query_to_reference(mdata_reference, mdata_query, adata_query)
+    else:
+        raise ValueError("Incorrect format of reference, please provide a .h5mu file")
+
+    adata_query.uns["integration_method"] = "totalvi"
+
+    logger.info("Getting the latent representation of query")
+    adata_query.obsm[par["obsm_output"]] = vae_query.get_latent_representation()
```

Comment:
Perhaps it is a good idea to save the output to a fresh copy of the input mudata? This would avoid any preparations of the input data to be compatible with totalVI of corrupting the output data?

Reply:
This is possible but would require copying everything from the input mudata with the risk of losing something or corruption anyway. If you think, that this would be cleaner than the current approach, I can do it :) What do you mean by corrupting the output by the way?

Reply:
I meant that the input file is being read and processed, and the new information is added to it.
If durin processing, some method accidentally adjusts something in-place, the original mudata file will be edited and passed to the output.

I think it is safer to either send a copy through the processing or add the output to a fresh copy of the original data. WDYT?

Reply:
Wait, but does reading `.h5mu` file, editing mudata object, and then saving it to new file change the input file? I ran a simple test, and looks like it is not the case:
```
mdata = mudata.read("../data/output.h5mu")
print(mdata.uns) # {}
mdata.uns["new_key"] = "value"
print(mdata.uns)  # {'new_key': 'value'}
mdata.write_h5mu("../data/new_output.h5mu")
mdata = mudata.read("../data/output.h5mu")
print(mdata.uns) # {}
```

Reply:
Discussed at the meeting: I will create a new AnnData object in `convert_mudata_to_anndata()`, containing only what I need for TOTALVI